internal/controller/cloudsyncmgr/cloudsync.go

```diff
@@ -65,37 +65,50 @@ func (c *CloudSyncImpl) InitiateCloudSync(isCloudSet string) (err error) {
 		if strings.Compare(strings.ToLower(isCloudSet), "true") == 0 {
 			log.Println("CloudSync init set")
 			isCloudSyncSet = true
+			//Intialize the client and hashmap storing client data
+			mqttmgr.SetClient(nil)
+			mqttmgr.InitClientData()
 		}
 	}
 	return nil
 }
 
-//StartCloudSync is used to start the sync by connecting to the broker
-func (c *CloudSyncImpl) StartCloudSync(host string) (err error) {
-	if isCloudSyncSet && len(host) > 0 {
-		log.Info("Starting the CLoudsync Mgr")
-		go mqttmgr.StartMQTTClient(host)
-	}
-	return
-}
-
 // RequestCloudSyncConf is  configuration request handler
-func (c *CloudSyncImpl) RequestCloudSyncConf(message mqttmgr.Message, topic string, clientID string) string {
+func (c *CloudSyncImpl) RequestCloudSyncConf(host string, clientID string, message mqttmgr.Message, topic string) string {
 	log.Info(logPrefix, "Publishing the data to the cloud")
-	mqttClient = mqttmgr.GetClient()
-	mqttmgr.SetClientID(clientID)
 	resp := ""
-	if mqttClient.IsConnected() {
+	var wg sync.WaitGroup
```

Comment:
What about implementing a routing for publishing a message like [detectNetworkChgRoutine()](https://github.com/lf-edge/edge-home-orchestration-go/blob/565cc43ac637dcd22ffc0d9206584103632fb5a4/internal/controller/discoverymgr/discovery.go#L124) instead of `WaitGroup`?

Reply:
This waitgroup is used instead of sleep to allow completion of go routine its not just related to network so for me it is correctly used. Thanks

Reply:
WaitGroup is synchronous. I recommend you to implement this asynchronously.

Reply:
But in this case, making it asynchronous wwould not serve the purpose. In the scenario of timeout, sleep was giving an error because the wait time was less than the retry time of mosquitto.connect. In such a scenario empty message was given to the user. Hence a synchronous wwait grp is needed so that the connect result is obtained if error or success and then only further stmts are executed.

Reply:
Make sense. I thought this should be handled asynchronously and recommended you to implement it via goroutine and channel. If this should be run synchronously, it must be good to get rid of goroutine, waitgroup and channel. Sorry for your inconvenience.

Reply:
go routine is necessary because else threads would be needed to allow the client to be connected.but go routines are cheaper than threads. Hence for each connection instead of thread a go routine is used.so it will be needed